contracts/Governance.sol

```diff
@@ -0,0 +1,285 @@
+        _;
+    }
+
+    modifier onlyWhenProposalIsDecided(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.Decided, "Only when proposal in Decided state.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriod(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp <= proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period.");
+        _;
+    }
+
+    modifier onlyWhenPermitted(bytes32 _id) {
+        require(proposalList.isPermitted(_id), "Only when proposal is permitted.");
+        _;
+    }
+
+    modifier onlyWhenQuorumReached(bytes32 _id) {
+        require(proposals[_id].quorumToken.balanceOf(supportDeposit) >= getRequiredQuorum(_id), "Only when propsal has quorum.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriodPassed(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp > proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period passed.");
+        _;
+    }
+
+    modifier onlyWhenApproved(bytes32 _id) {
+        require(proposals[_id].approved, "Only when proposal is approved.");
+        _;
+    }
+
+
+    // ****************************** //
+    // *           Events           * //
+    // ****************************** //
+
+    event ProposalPutToSupport(bytes32 indexed _id);
+
+    event ProposalPutToVote(bytes32 indexed _id);
+
+    event ProposalDecided(bytes32 indexed _id, bool _approved);
+
+    event ProposalExecuted(bytes32 indexed _id);
+
+    event ProposalCreated(bytes32 indexed _id, address _destination, uint _amount, bytes _data, string _uriDescription, bytes32 _descriptionHash, string _uriArguments, bytes32 _argumentsHash);
+
+
+    // ****************************** //
+    // *    Governance Mechanism    * //
+    // ****************************** //
+
+    /** @dev Creates a proposal and adds to ProposalState
+     *  @param _id ID of the proposalList
+     *  @param _destination Destination contract of the execution
+     *  @param _amount Value of the execution
+     *  @param _data Data of the execution
+     *  @param _uriDescription URI of the description of the proposal
+     *  @param _uriArguments URI of the arguments of the proposal
+     */
+    function createProposal(bytes32 _id, address _destination, uint _amount, bytes _data, string _uriDescription, string _uriArguments) public {
+        require(proposals[_id].destination == address(0), "There is already a proposal with this id");
+
+        proposals[_id].destination = _destination;
+        proposals[_id].amount = _amount;
+        proposals[_id].data = _data;
+        proposals[_id].uriDescription = _uriDescription;
+        proposals[_id].uriArguments = _uriArguments;
+
+        proposals[_id].descriptionHash = keccak256(proposals[_id].uriDescription);
```

Comment:
These are supposed to be the hashes of the file contents, not the URIs, so they should be params to the function.

Reply:
I couldn't understand this, how to hash the content as we don't keep it on contract?

Reply:
The submitter hashes it off chain and submits it as a param to the contract. The contract can't verify but whoever looks up the proposal can